test/unit/codegen/codegen_helper.cpp

```diff
@@ -0,0 +1,130 @@
+ *************************************************************************/
+
+#define CATCH_CONFIG_MAIN
+
+#include <catch/catch.hpp>
+
+#include "ast/program.hpp"
+#include "codegen/codegen_helper_visitor.hpp"
+#include "parser/nmodl_driver.hpp"
+#include "visitors/symtab_visitor.hpp"
+
+using namespace nmodl;
+using namespace visitor;
+using namespace codegen;
+
+using nmodl::parser::NmodlDriver;
+
+//=============================================================================
+// Helper for codege related visitor
+//=============================================================================
+std::string run_inline_visitor(const std::string& text) {
+    NmodlDriver driver;
+    const auto& ast = driver.parse_string(text);
+
+    /// construct symbol table and run codegen helper visitor
+    SymtabVisitor().visit_program(*ast);
+    CodegenHelperVisitor v;
+
+    /// symbols/variables are collected in info object
+    const auto& info = v.analyze(*ast);
+
+    /// semicolon separated list of variables
+    std::string variables;
+
+    /// range variables in order of code generation
+    for (const auto& var: info.range_parameter_vars) {
+        variables += var->get_name() + ";";
+    }
+    for (const auto& var: info.range_assigned_vars) {
+        variables += var->get_name() + ";";
+    }
+    for (const auto& var: info.range_state_vars) {
+        variables += var->get_name() + ";";
+    }
+    for (const auto& var: info.assigned_vars) {
+        variables += var->get_name() + ";";
+    }
+
+    return variables;
+}
+
+SCENARIO("unusual / failing mod files", "[codegen][var_order]") {
+    GIVEN("cal_mig.mod : USEION variables declared as RANGE") {
+        std::string nmodl_text = R"(
+            PARAMETER {
+              gcalbar=.003 (mho/cm2)
+              ki=.001 (mM)
+              cai = 50.e-6 (mM)
+              cao = 2 (mM)
+              q10 = 5
+              USEGHK=1
+            }
+            NEURON {
+              SUFFIX cal
+              USEION ca READ cai,cao WRITE ica
+              RANGE gcalbar, cai, ica, gcal, ggk
+              RANGE minf, tau
+              GLOBAL USEGHK
+            }
+            STATE {
+              m
+            }
+            ASSIGNED {
+              ica (mA/cm2)
+              gcal (mho/cm2)
+              minf
+              tau   (ms)
+              ggk
+            }
+            DERIVATIVE state {
+              rate(v)
+              m' = (minf - m)/tau
+            }
+        )";
+
+        THEN("ionic current variable declared as RANGE appears first") {
+            std::string expected = "gcalbar;ica;gcal;minf;tau;ggk;m;cai;cao;";
+            auto result = run_inline_visitor(nmodl_text);
+            REQUIRE(result == expected);
+        }
+    }
+
+    GIVEN("CaDynamics_E2.mod : USEION variables declared as STATE variable") {
+        std::string nmodl_text = R"(
+            NEURON  {
+              SUFFIX CaDynamics_E2
+              USEION ca READ ica WRITE cai
```

Comment:
I think that there are some mod files that have the same variable in `READ` and `WRITE`? What happens with those ones? Maybe a test would be useful?
Looking at the code I think that they are handled the same way. The only issue might be that they are declared 2 times?

Reply:
cadyn.mod unit test added. Its handled correctly.